### PR TITLE
DXF multilayer support

### LINKED ISSUE #1171: dxf export: support layers
ezdxf [supports dxf layers](https://ezdxf.readthedocs.io/en/stable/concepts/layers.html).

For instance, this code generates a dxf file with 256 layers, one for each available color:

```py
import ezdxf
import random

doc = ezdxf.new()
msp = doc.modelspace()

for color in range(256):
    layer_name = f'layer{ color }'
    doc.layers.add(name=layer_name, color=color)

    start = random.randint(0, 100), random.randint(0, 100)
    end = random.randint(0, 100), random.randint(0, 100)
    msp.add_line(start, end, dxfattribs={'layer': layer_name})

doc.saveas('test.dxf')
```

Result (after importing the dxf with Inkscape):
image

Multi-layers dxf can be used with cnc or laser cutters in order to specify which tool or options must be used for each (ex: one layer to engrave, another one to cut).

Each layers supports:
- a color: which is a number between 0 and 255, corresponding to an index in the [color palette](https://ezdxf.readthedocs.io/en/stable/concepts/aci.html#aci);
- a [line type](https://ezdxf.readthedocs.io/en/stable/concepts/linetypes.html) (continuous, dashed, etc.);
- a [line weight](https://ezdxf.readthedocs.io/en/stable/concepts/lineweights.html), which is a value between 0 and 211 where 1mm = 100 units.

Those options could be passed in the `opt` export attribute:

```py
cq.exporters.export(shape, 'test.dxf', opt={
    'color': 5,
    'line_type': 'dashed',
    'line_weight': 25
})
```

For multi-layers export, I'm not sure what is the best syntax. I have several approaches in mind:

1. pass lists for each option:

```py
cq.exporters.export([shape1, shape2, shape3], 'test.dxf', opt={
    'color': [5, 6, 7],
    'line_type': ['continuous', 'dot', 'dashed'],
    'line_weight': [18, 25, 40]
})
```

2. Add a new class dedicated to DXF export:

```py
layer1 = DxfLayer(shape1, 5, DxfLayer.LineType.CONTINUOUS, 18)
layer2 = DxfLayer(shape2, 6, DxfLayer.LineType.DOT, 25)
layer3 = DxfLayer(shape3, 7, DxfLayer.LineType.DASHED, 40)
cq.exporters.export([ layer1, layer2, layer3 ], 'test.dxf')
```

3. Add 2 classes dedicated to DXF export, one for the layer and one for the model:

```py
layer1 = DxfLayer('layer1', 5, DxfLayer.LineType.CONTINUOUS, 18)
layer2 = DxfLayer('layer2', 6, DxfLayer.LineType.DOT, 25)
layer3 = DxfLayer('layer3', 7, DxfLayer.LineType.DASHED, 40)

dxf_model = (DxfModel()
    .add(shape1, layer1)
    .add(shape2, layer2)
    .add(shape3, layer3)
)

cq.exporters.export(dxf_model, 'test.dxf')
```

I would like to work on this, since I need this for my project, so any advice on the export syntax will be welcome.

### LINKED ISSUE #1188: dxf export: support setting units
The `exportDXF` function should take an `opts` arguments like the other exports that allows setting the [intended modelspace drawing units](https://ezdxf.mozman.at/docs/concepts/units.html#insunits) via the `$INSUNITS` header value.

[Since v0.15 ezdxf provides the common units](https://github.com/mozman/ezdxf/commit/eda3efb98783a3242964cf70763836f70a081068#diff-8c4c0b9e88521c52fa84e802ca99b0b4bc48c899ce2ae55de196cd694e35c833R19), so there could be a `units: 'cm'` value in the `opts` argument, and applied something like this:

```python
if "units" in opts:
  units = opts["units"]
  if isinstance(units, str):
    units = getattr(ezdxf.units, units)
  doc.header['$INSUNITS'] = units
```

### LINKED ISSUE #1259: Export step/stl in milimeter, export dxf is meter
Hi, is there an easy solution for my problem that in standard step/stl export is in mm and dxf is in m?
```
import cadquery as cq
from cadquery import exporters
import cadquery as cq
import math
box1 = cq.Workplane().box(10,20,30)
exporters.export(box1, './box1.step')
box1Face = box1.faces(">Z")
exporters.export(box1Face, './box1Face.dxf')
```
It's about the parameter $INSUNITS$ in dxf-file.
value 6 (meter) is exported, value 4 is for milimeters and would be nice.

I've tried to change parameter with an option
```
exporters.export(
    box1Face,
    './box1Face.dxf',
    opt={"$INSUNITS$":4,
    },
)
```
but did not work.
Any easy solution except for manipulating the dxf-code afterwards?

Thank you
Markus

### PR DESCRIPTION
**Pull request by sethfischer**
A proposed approach to #1171.

Also supports:

* units (see #1188, #1259)
* setting color by layer
* setting line type by layer

Closes #1171
Closes #1188
Closes #1259

## Todo

* [X] documentation
* [X] tests

## Notes

* The [ezdxf](https://ezdxf.readthedocs.io/) document object is available via the public API so most features of ezdxf are accessible.
* `cadquery.occ_impl.exporters.dxf.exportDXF` has be refactored but there is **no change** to the public API.
* Methods `_dxf_*` have been refactored to better facilitate unit testing. There is **no change** to algorithm of these methods.

## Example use

```python
"""DXF multilayer demo."""

import cadquery as cq
from cadquery.occ_impl.exporters.dxf import DxfDocument

spline_points = [(90, 50), (83, 60), (65, 50), (50, 33), (33, 40), (15, 35), (0, 35)]
spline = (
    cq.Workplane("XY")
    .lineTo(100, 0)
    .lineTo(100, 35)
    .spline(spline_points, includeCurrent=True)
    .close()
)

circle = cq.Workplane().moveTo(15, 15).circle(10)
ellipse = cq.Workplane().moveTo(45, 15).ellipse(15, 10, 0)
rectangle = cq.Workplane().moveTo(70, 15).rect(10, 20)
fold_line = cq.Workplane().moveTo(80, -5).line(0, 70)

# create the DXF document

dxf = DxfDocument(setup=True)

dxf = (
    dxf.add_layer("layer_cut", color=2)
    .add_layer("layer_line_engrave", color=3)
    .add_layer("layer_area_engrave", color=4)
    .add_layer("layer_fold", color=5, linetype="DASHED")
    .add_shape(spline, "layer_cut")
    .add_shape(ellipse, "layer_cut")
    .add_shape(circle, "layer_line_engrave")
    .add_shape(rectangle, "layer_area_engrave")
    .add_shape(fold_line, "layer_fold")
)

dxf.document.saveas("multilayer-demo.dxf")
```

Screenshot of resulting `multilayer-demo.dxf` viewed in LibreCAD:

1267-librecad-screenshot_2023-02-10

This still works as expected but is refactored to call `DxfDocument` and **DXF document units are mm**:

```python
import cadquery as cq
from cadquery import exporters

result = cq.Workplane().box(10, 10, 10)

exporters.export(result, "export.dxf")
```

Export DXF with document units other then the default:

```python
exporters.export(
    result,
    "export.dxf",
    opt={"doc_units": 6},  # set DXF document units to meters
)

exporters.exportDXF(
    result,
    "export.dxf",
    doc_units=6,  # set DXF document units to meters
)
```

### CODE REVIEW

**Comment by adam-urbanczyk** [comment]
You might to want to wait with this until #1226  is merged.

**Comment by sethfischer** [comment]
> You might to want to wait with this until #1226 is merged.

@adam-urbanczyk Will do. Once merged I'll rebase this work on master

**Comment by sethfischer** [comment]
I've rebased this on master pulling in #1226 which introduced parameters `approx` and `tolerance`.

Should `approx` and `tolerance` be:

1. properties of the DXF document? or,
2. properties of the shape being added?, or
3. properties of both the DXF document and the shape?


## 1. Properties of the DXF document

Properties of the DXF document applying to all shapes added with `add_shape`.

```python
dxf = DxfDocument(approx: Optional[Literal["spline", "arc"]] = None, tolerance: float = 1e-3)
dxf.add_shape(shape)
```

## 2. Properties of the shape being added

Property of the shape added with `add_shape`.

```python
dxf = DxfDocument()
dxf.add_shape(shape, approx: Optional[Literal["spline", "arc"]] = None, tolerance: float = 1e-3)
```

## 3. Properties of both the DXF document and the shape

Properties of the DXF document applying to all shapes added with `add_shape`. But specific shapes can have their own `approx` and `tolerance` values.

```python
dxf = DxfDocument(approx: Optional[Literal["spline", "arc"]] = None, tolerance: float = 1e-3)
dxf.add_shape(shape, approx: Optional[Literal["spline", "arc"]] = None, tolerance: float = 1e-3)
```

**Comment by adam-urbanczyk** [comment]
Definitely a property of the document. You might also add an option to override on per-shape basis, but I personally don't have such a use case.

File: cadquery/occ_impl/exporters/dxf.py [hunk 1]
```diff
@@ -1,154 +1,328 @@
...
+from ...cq import Face, Plane, Workplane
+from ...units import RAD2DEG
+from ..shapes import Edge
+from .utils import toCompound
 
-    c = ellipse.Location()
-    xdir = ellipse.XAxis().Direction()
-    xax = r2 * xdir.XYZ()
+ApproxOptions = Literal["spline", "arc"]
+DxfEntityAttributes = Tuple[str, Dict[str, Any]]
```
Comment (adam-urbanczyk) [inline cadquery/occ_impl/exporters/dxf.py:17]: Could you type the first `str` using `Literal['SPLINE','ARC','CIRCLE','ELLIPSE']`

File: cadquery/occ_impl/exporters/dxf.py [hunk 2]
```diff
@@ -1,154 +1,328 @@
...
+
+        self.document = ezdxf.new(dxfversion=dxfversion, setup=setup, units=doc_units)  # type: ignore[attr-defined]
+        self.msp = self.document.modelspace()
+
+        doc_metadata = self.document.ezdxf_metadata()
+        for key, value in metadata.items():
+            doc_metadata[key] = value
+
+    def add_layer(
+        self, name: str, *, color: int = 1, linetype: str = "CONTINUOUS"
```
Comment (adam-urbanczyk) [inline cadquery/occ_impl/exporters/dxf.py:108]: Could you type linetype using Literal[...]?

**Review by adam-urbanczyk — CHANGES_REQUESTED** [review]
LGTM, small changes

File: cadquery/occ_impl/exporters/dxf.py [hunk 3]
```diff
@@ -1,154 +1,328 @@
...
+    def __init__(
+        self,
+        dxfversion: str = "AC1027",
+        setup: Union[bool, List[str]] = False,
+        doc_units: int = units.MM,
+        *,
+        metadata: Union[Dict[str, str], None] = None,
+        approx: Optional[ApproxOptions] = None,
+        tolerance: float = 1e-3,
+    ) -> None:
```
Comment (adam-urbanczyk) [inline cadquery/occ_impl/exporters/dxf.py:67]: AFAIR `-> None` is not needed (https://github.com/python/mypy/issues/604)

File: cadquery/occ_impl/exporters/dxf.py [hunk 4]
```diff
@@ -1,154 +1,328 @@
...
+        self.document = ezdxf.new(dxfversion=dxfversion, setup=setup, units=doc_units)  # type: ignore[attr-defined]
+        self.msp = self.document.modelspace()
+
+        doc_metadata = self.document.ezdxf_metadata()
+        for key, value in metadata.items():
+            doc_metadata[key] = value
+
+    def add_layer(
+        self, name: str, *, color: int = 1, linetype: str = "CONTINUOUS"
+    ) -> "DxfDocument":
```
Comment (adam-urbanczyk) [inline cadquery/occ_impl/exporters/dxf.py:109]: I'd propose to use `Self` here. `typing_extensions` should have it.

File: cadquery/occ_impl/exporters/dxf.py [hunk 5]
```diff
@@ -1,154 +1,328 @@
...
+
+        :param name: layer definition name
+        :param color: ezdxf color
+        :param linetype: ezdxf :doc:`line type <ezdxf-stable:concepts/linetypes>`
+        """
+        self.document.layers.add(name, color=color, linetype=linetype)
+
+        return self
+
+    def add_shape(self, workplane: Workplane, layer: str = "") -> "DxfDocument":
```
Comment (adam-urbanczyk) [inline cadquery/occ_impl/exporters/dxf.py:123]: Self again

**Comment by adam-urbanczyk** [comment]
@lorenzncode you might want to integrate #1278 into this PR

File: cadquery/occ_impl/exporters/dxf.py (hunk 2 above)
Comment (sethfischer) [reply cadquery/occ_impl/exporters/dxf.py:111]: The value of `linetype` could be any string the user chooses.

Ezdxf creates several [standard linetypes](https://ezdxf.readthedocs.io/en/stable/concepts/linetypes.html) which are supported by all DXF versions, but it is also possible for the user to [create custom line types](https://ezdxf.readthedocs.io/en/stable/tutorials/linetypes.html).

Via the `__init__` `setup=True` parameter a user may choose not to set up standard linetypes.

From https://ezdxf.readthedocs.io/en/stable/concepts/linetypes.html :

> CONTINUOUS is the default linetype for layers with an unspecified linetype.

**Comment by sethfischer** [comment]
@adam-urbanczyk back to you for review.

File: cadquery/occ_impl/exporters/dxf.py (hunk 2 above)
Comment (adam-urbanczyk) [reply cadquery/occ_impl/exporters/dxf.py:111]: Clear, thanks!

File: cadquery/occ_impl/exporters/dxf.py [hunk 6]
```diff
@@ -1,154 +1,331 @@
...
+
+        self.approx = approx
+        self.tolerance = tolerance
+
+        self.document = ezdxf.new(dxfversion=dxfversion, setup=setup, units=doc_units)  # type: ignore[attr-defined]
+        self.msp = self.document.modelspace()
+
+        doc_metadata = self.document.ezdxf_metadata()
+        for key, value in metadata.items():
+            doc_metadata[key] = value
```
Comment (adam-urbanczyk) [inline cadquery/occ_impl/exporters/dxf.py:108]: This line seems to be not covered

File: cadquery/occ_impl/exporters/dxf.py [hunk 7]
```diff
@@ -1,154 +1,331 @@
...
+
+        :param workplane: CadQuery Workplane
+        :param layer: layer definition name
+        """
+        plane = workplane.plane
+        shape = toCompound(workplane).transformShape(plane.fG)
+
+        general_attributes = {}
+        if layer:
+            general_attributes["layer"] = layer
```
Comment (adam-urbanczyk) [inline cadquery/occ_impl/exporters/dxf.py:137]: Same here: not covered

https://app.codecov.io/gh/CadQuery/cadquery/pull/1267/blob/cadquery/occ_impl/exporters/dxf.py

**Review by adam-urbanczyk — APPROVED** [review]
LGTM, last two remarks regarding diff coverage. Could you extend the tests?

**Review by adam-urbanczyk — APPROVED** [review]

File: doc/importexport.rst [hunk 8]
```diff
@@ -239,17 +239,86 @@ optimum values that will produce an acceptable mesh.
...
+    # or
+    exporters.export(result, "/path/to/file/object.dxf")
+
+
+Units
+-----
+
+The default DXF document units are mm (:code:`doc_units = 4`), however the document units can be set to any
+:doc:`unit supported by ezdxf <ezdxf-stable:concepts/units>`.
+
```
Comment (lorenzncode) [inline doc/importexport.rst:284]: @sethfischer How about including a quick reference of common units directly say 0 to 6?
```
The default DXF document units are mm (:code:`doc_units = 4`).

========= ===============
doc_units Unit
========= ===============
0         Unitless
1         Inches
2         Feet
3         Miles
4         Millimeters
5         Centimeters
6         Meters
========= ===============

Document units can be set to any :doc:`unit supported by ezdxf <ezdxf-stable:concepts/units>`.
```

File: doc/importexport.rst [hunk 9]
```diff
@@ -239,17 +239,86 @@ optimum values that will produce an acceptable mesh.
...
+        ``"spline"``
+            all splines approximated as cubic splines
+        ``"arc"``
+            all curves approximated as arcs and straight segments
+``tolerance``
+    Approximation tolerance for converting :class:`cadquery.Workplane` objects to DXF entities.
+    See `Approximation strategy`_.
+``doc_units``
+    Ezdxf document/modelspace :doc:`units <ezdxf-stable:concepts/units>`.
+    See `Approximation strategy`_.
```
Comment (lorenzncode) [inline doc/importexport.rst:264]: approx link is repeated. Change:
See `Units`_.

File: cadquery/occ_impl/exporters/dxf.py [hunk 10]
```diff
@@ -1,154 +1,331 @@
...
 
-    msp.add_ellipse(
-        (c.X(), c.Y(), c.Z()),
-        (xax.X(), xax.Y(), xax.Z()),
-        r1 / r2,
-        geom.FirstParameter(),
-        geom.LastParameter(),
-    )
 
+class DxfDocument:
```
Comment (lorenzncode) [inline cadquery/occ_impl/exporters/dxf.py:23]: Add `DxfDocument` to `exporters/__init__.py`?:
```
from .dxf import exportDXF, DxfDocument
```

Then usage becomes;
```py
import cadquery as cq

dxf = cq.exporters.DxfDocument()
```

**Comment by lorenzncode** [comment]
I've added the fix from #1278.  Thanks @sethfischer!

**Review by lorenzncode — APPROVED** [review]

**Review by adam-urbanczyk — APPROVED** [review]

**Comment by adam-urbanczyk** [comment]
Thanks @sethfischer

**Comment by sethfischer** [comment]
Thank you @adam-urbanczyk and @lorenzncode

**Pull request merged**